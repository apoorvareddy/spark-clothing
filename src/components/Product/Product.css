.row{
  justify-content: space-evenly;
}
.card .card-img-top {
  max-width: 260px;
  max-height: 280px;
  margin: 0 auto;
}
.card .card-body {
  text-align: center;
}
.card {
  margin-top: 15px;
  border: none;
}
.col-12.col-sm-6.col-md-4.card {
  border: none;
}
.card-content{
  border: 1px solid rgb(51, 50, 50);
  border-radius: 5px;
  text-align: center;
}
.tag-line{
  font-size: 12px;
  margin-top: 10px;
  font-weight: normal;
}
.view-all-button{
  width: 150px;
  margin-top: 20px;
}
.card-title a {
  color: #333;
  font-weight: bold;
}
