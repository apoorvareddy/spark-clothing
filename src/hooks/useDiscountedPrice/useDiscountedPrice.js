// getting the MRP and the discount and calculating the discounted price
const useDiscountedPrice = (originalPrice, discountPercentage) => {
  const discountedPrice = Math.round(originalPrice - (originalPrice * discountPercentage) / 100);
  return { originalPrice, discountedPrice }
}

export default useDiscountedPrice;
