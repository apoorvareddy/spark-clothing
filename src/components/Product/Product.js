import Button from 'react-bootstrap/Button';
import Card from 'react-bootstrap/Card';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Product.css';
import useDiscountedPrice from '../../hooks/useDiscountedPrice/useDiscountedPrice';

const Product = ({ imageUrl, id, name, maxRetailPrice, tagLine, discountApplicable, imgAltText }) => {
  // using useDiscountedPrice hook for discountedPrice
  const { discountedPrice } = useDiscountedPrice(maxRetailPrice, discountApplicable);
  return (
    <Card className='col-12 col-sm-6 col-md-4'>
      <div className='card-content'>
        <Card.Img variant="top" data-testid='imageUrl' src={imageUrl} alt={imgAltText} />

        <Card.Body>
          <Card.Title>
            <Link to={`/products/${id}`} data-testid='name'>{name}</Link>
            <p className='tag-line' data-testid='tagLine'>{tagLine}</p>
          </Card.Title>

          <Card.Text>
            <span className='discount-price' data-testid='discountedPrice'>Rs.{discountedPrice}</span>
            <s data-testid='maxRetailPrice'>Rs.{maxRetailPrice}</s>
            <span className='discount-percentage' data-testid='discountApplicable'>({discountApplicable}% OFF)</span>
          </Card.Text>
          <Button variant="secondary">Add to Cart</Button>
        </Card.Body>
      </div>
    </Card>
  )
}

Product.propTypes = {
  imageUrl: PropTypes.string,
  id: PropTypes.number,
  name: PropTypes.string,
  maxRetailPrice: PropTypes.number,
  tagLine: PropTypes.string,
  discountApplicable: PropTypes.number,
  imgAltText: PropTypes.string
}

export default Product;
