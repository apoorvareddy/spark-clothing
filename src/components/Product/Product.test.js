import { render, screen } from '@testing-library/react';
import { HashRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
// import { renderHook } from '@testing-library/react-hooks';
import useDiscountedPrice from '../../hooks/useDiscountedPrice/useDiscountedPrice';
import Product from './Product';

describe('Product', () => {
  // checking Product component renders the correct discounted price using useDiscountedPrice hook
  it('receives maxRetailPrice and discountApplicable and calculates the discountedPrice', () => {
    const maxRetailPrice = 100;
    const discountApplicable = 20;
    render(
      <HashRouter>
        <Product maxRetailPrice={maxRetailPrice} discountApplicable={discountApplicable} />
      </HashRouter>
    );

    const { originalPrice, discountedPrice } = useDiscountedPrice(maxRetailPrice, discountApplicable);
    expect(originalPrice).toBe(100);
    expect(discountedPrice).toBe(80);
  });

  it('renders all the product information correctly', () => {
    render(
      <HashRouter>
        <Product
          id={1}
          name='U.S. Polo'
          maxRetailPrice={5000}
          tagLine='white shirt'
          discountApplicable={10}
        />
      </HashRouter>
    )

    expect(screen.getByText('U.S. Polo')).toBeInTheDocument();
    expect(screen.getByText('Rs.5000')).toBeInTheDocument();
  })

  it('have right snapshot with all the requirements completed', () => {
    const snapshotInJson = renderer.create(
      <HashRouter>
        <Product
          imageUrl='images/ucb-t-shirt.jpg'
          name='UCB'
          tagLine='Full Sleeves Sweatshirt Text Print - Blue'
          imgAltText='UCB Full Sleeves Sweatshirt Text Print - Blue'
        />
      </HashRouter>
    ).toJSON();
    expect(snapshotInJson).toMatchSnapshot();
  })
});
