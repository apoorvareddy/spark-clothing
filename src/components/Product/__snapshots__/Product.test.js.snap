// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product have right snapshot with all the requirements completed 1`] = `
<div
  className="col-12 col-sm-6 col-md-4 card"
>
  <div
    className="card-content"
  >
    <img
      alt="UCB Full Sleeves Sweatshirt Text Print - Blue"
      className="card-img-top"
      data-testid="imageUrl"
      src="images/ucb-t-shirt.jpg"
    />
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        <a
          data-testid="name"
          href="#/products/undefined"
          onClick={[Function]}
        >
          UCB
        </a>
        <p
          className="tag-line"
          data-testid="tagLine"
        >
          Full Sleeves Sweatshirt Text Print - Blue
        </p>
      </div>
      <p
        className="card-text"
      >
        <span
          className="discount-price"
          data-testid="discountedPrice"
        >
          Rs.
          NaN
        </span>
        <s
          data-testid="maxRetailPrice"
        >
          Rs.
        </s>
        <span
          className="discount-percentage"
          data-testid="discountApplicable"
        >
          (
          % OFF)
        </span>
      </p>
      <button
        className="btn btn-secondary"
        disabled={false}
        type="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
</div>
`;
